src/H5system.c

```diff
@@ -935,8 +935,170 @@ H5_expand_windows_env_vars(char **env_var)
+
+    len = HDstrlen(s);
+
+    if (len > n)
+        len = n;
+
+    if (NULL == (ret_value = HDmalloc(len + 1)))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, NULL, "can't allocate buffer for string")
+
+    HDmemcpy(ret_value, s, len);
+    ret_value[len] = '\0';
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+}
 #endif /* H5_HAVE_WIN32_API */
 
+/* dirname() and basename() are not easily ported to Windows and basename
+ * behavior varies depending on if you get POSIX vs. GNU. As a more
+ * platform-indpendent work-around, we've implemented H5_ versions of
+ * dirname() and basename().
+ *
+ * - The input string is never modified.
+ *
+ * - The out parameter is a new string that was allocated with HDmalloc()
+ *   and must be freed by the caller via HDfree().
+ *
+ * - NULL pointers are errors.
+ *
+ * - On errors, FAIL will be returned and the output parameter will be
+ *   undefined.
+ *
+ * - The trailing file separator is not returned in H5_dirname().
+ *
+ * - Passing a path that ends in a file separator to H5_basename() will
+ *   return the empty string.
+ *
+ * - If the path doesn't contain a file separator, we assume you passed
+ *   in a raw filename and H5_basename() will return the input string
+ *   while H5_dirname will return ".".
+ *
+ * - Assumes the file separator is \ on Win32 and / everywhere else,
+ *   including Cygwin.
+ */
+
+herr_t
+H5_dirname(const char *path, char **dirname)
+{
+    const char *sep;
+    char       *out       = NULL;
+    herr_t      ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI_NOINIT
+
+    if (!path)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "path can't be NULL")
+    if (!dirname)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "dirname can't be NULL")
+
+    if (NULL == (sep = HDstrrchr(path, H5_DIR_SEPC))) {
+        out = HDstrdup(".");
+    }
+    else if (sep == path) {
+        out = HDstrdup(H5_DIR_SEPS);
+    }
+    else {
+        ptrdiff_t len;
+
+        if (sep[1] == '\0') {
+            /* Find a previous occurrence of the file separator, if any */
+            while (sep != path) {
+                sep--;
+                if (*sep == H5_DIR_SEPC)
+                    break;
+            }
+
+            if (sep == path && *sep != H5_DIR_SEPC) {
+                out = HDstrdup(".");
+            }
+            else {
+                len = (sep - path) + 1;
+
+                HDassert(len >= 0);
+
+                out = HDstrndup(path, (size_t)len);
+            }
+        }
+        else {
+            len = sep - path;
+
+            HDassert(len >= 0);
+
+            out = HDstrndup(path, (size_t)len);
+        }
+    }
+
+    if (NULL == out)
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "can't allocate buffer for dirname");
+
+    *dirname = out;
+
+done:
+    if (FAIL == ret_value) {
+        HDfree(out);
+        if (dirname)
+            *dirname = NULL;
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5_dirname() */
+
+herr_t
```

Comment:
I see that there's discussion about basename in the comment above dirname, but if a developer is looking at basename <n> years from now, they probably won't see it.  Helpful comments here should be added.

Reply:
Added a minimal header to each one, presuming that developers know how to `man dirname`/`man basename` for any other relevant details.

Reply:
Looks good